### ISSUE TITLE
PostgreSQL "column expansion" causes a "templating/parsing error"

### ISSUE THREAD

**Issue by dhermes** [sqlfluff/sqlfluff#6811]
### Search before asking

- [x] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

For the following minimal example:

```sql
-- before.sql
SELECT
    w.id AS widget_id,
    (JSONB_EACH_TEXT(w.inventory_events)).*
FROM public.widget AS w
WHERE w.id IN (6023, 43763, 299888);
```

a parsing failure occurs:

```
$ python -m sqlfluff fix --dialect postgres before.sql
==== finding fixable violations ====
  [1 templating/parsing errors found]                                                                                                                                                                          
==== no fixable linting violations found ====
All Finished 📜 🎉!
$
$
$ python -V
Python 3.12.7
$
$
$ python -m sqlfluff --version
python -m sqlfluff, version 3.3.1
```

The `(JSONB_EACH_TEXT(w.inventory_events)).*` is a PostgreSQL "column expansion", which is syntactic sugar that essentially says "splat all the columns within the current selector".

### Expected Behaviour

SQL should be parse-able. The

```sql
-- after.sql
SELECT
    w.id AS widget_id,
    w_event.key,
    w_event.value
FROM public.widget AS w
LEFT JOIN
    LATERAL JSONB_EACH_TEXT(w.inventory_events) AS w_event (key, value)
    ON TRUE
WHERE w.id IN (6023, 43763, 299888);
```

```
$ python -m sqlfluff fix --dialect postgres after.sql
==== finding fixable violations ====
==== no fixable linting violations found ====                                                                                                                                                                  
All Finished 📜 🎉!
```

Note that the `LEFT JOIN LATERAL` form is recommended: https://www.postgresql.org/docs/current/rowtypes.html

> PostgreSQL handles column expansion by actually transforming the first form into the second. So, in this example, `myfunc()` would get invoked three times per row with either syntax. If it's an expensive function you may wish to avoid that, which you can do with a query like:
>
> ```sql
> SELECT m.* FROM some_table, LATERAL myfunc(x) AS m;
> ```
>
> Placing the function in a `LATERAL FROM` item keeps it from being invoked more than once per row. `m.*` is still expanded into `m.a`, `m.b`, `m.c`, but now those variables are just references to the output of the `FROM` item. (The `LATERAL` keyword is optional here, but we show it to clarify that the function is getting `x` from `some_table`.)

And to verify, both forms produce the same data:

```
acme> \i before.sql
+-----------+----------------------------------+-------+
| widget_id | key                              | value |
|-----------+----------------------------------+-------|
| 6023      | 2021-09-30T01:35:09.587643-08:00 | 4     |
| 6023      | 2023-12-16T12:29:33.182813-08:00 | 2     |
| 43763     | 2022-11-11T00:45:49.902560-07:00 | 2     |
| 299888    | 2023-03-16T21:37:27.593596-07:00 | 3     |
| 299888    | 2023-11-15T15:25:22.590514-07:00 | 3     |
| 299888    | 2024-02-18T02:20:36.682245-07:00 | 3     |
+-----------+----------------------------------+-------+
SELECT 6
Time: 0.086s

acme> \i after.sql
+-----------+----------------------------------+-------+
| widget_id | key                              | value |
|-----------+----------------------------------+-------|
| 6023      | 2021-09-30T01:35:09.587643-08:00 | 4     |
| 6023      | 2023-12-16T12:29:33.182813-08:00 | 2     |
| 43763     | 2022-11-11T00:45:49.902560-07:00 | 2     |
| 299888    | 2023-03-16T21:37:27.593596-07:00 | 3     |
| 299888    | 2023-11-15T15:25:22.590514-07:00 | 3     |
| 299888    | 2024-02-18T02:20:36.682245-07:00 | 3     |
+-----------+----------------------------------+-------+
SELECT 6
Time: 0.088s
```

### Observed Behaviour

```
$ python -m sqlfluff fix --dialect postgres before.sql
==== finding fixable violations ====
  [1 templating/parsing errors found]                                                                                                                                                                          
==== no fixable linting violations found ====
All Finished 📜 🎉!
```

I also expected to see the line / column where the parsing failure happened.

### How to reproduce

```sql
-- before.sql
SELECT
    w.id AS widget_id,
    (JSONB_EACH_TEXT(w.inventory_events)).*
FROM public.widget AS w
WHERE w.id IN (6023, 43763, 299888);
```

If you like, here are some dummy tables with the data from above:

```sql
BEGIN;

CREATE TABLE public.widget (
    id INTEGER PRIMARY KEY, inventory_events JSONB NOT NULL
);

INSERT INTO
public.widget (id, inventory_events)
VALUES
(
    6023,
    JSONB_BUILD_OBJECT(
        '2021-09-30T01:35:09.587643-08:00',
        4,
        '2023-12-16T12:29:33.182813-08:00',
        2
    )
),
(43763, JSONB_BUILD_OBJECT('2022-11-11T00:45:49.902560-07:00', 2)),
(
    299888,
    JSONB_BUILD_OBJECT(
        '2023-03-16T21:37:27.593596-07:00',
        3,
        '2023-11-15T15:25:22.590514-07:00',
        3,
        '2024-02-18T02:20:36.682245-07:00',
        3
    )
);

SELECT
    w.id AS widget_id,
    w_event.key,
    w_event.value
FROM public.widget AS w
LEFT JOIN
    LATERAL JSONB_EACH_TEXT(w.inventory_events) AS w_event (key, value)
    ON TRUE
WHERE w.id IN (6023, 43763, 299888);

ROLLBACK;
```

### Dialect

PostgreSQL

### Version

3.3.1

### Configuration

Default! Just the `--dialect` flag

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
